server/repository/src/migrations/v1_02_00/context_table.rs

```diff
@@ -0,0 +1,178 @@
+use crate::migrations::sql;
+use crate::{
+    ContextRow, ContextRowRepository, ProgramFilter, ProgramRepository, ProgramRowRepository,
+    StorageConnection, PATIENT_CONTEXT_ID,
+};
+
+pub(crate) fn migrate(connection: &StorageConnection) -> anyhow::Result<()> {
+    sql!(
+        connection,
+        r#"
+        CREATE TABLE context (
+          id TEXT NOT NULL PRIMARY KEY,
+          name TEXT NOT NULL
+        );
+        "#
+    )?;
+
+    #[cfg(not(feature = "postgres"))]
+    sql!(
+        connection,
+        r#"
+        PRAGMA foreign_keys = 0;
+        ALTER TABLE program ADD COLUMN context_id TEXT NOT NULL DEFAULT temp REFERENCES context(id);
+        PRAGMA foreign_keys = 1;
+        "#
+    )?;
+    #[cfg(feature = "postgres")]
+    sql!(
+        connection,
+        r#"
+        ALTER TABLE program ADD COLUMN context_id TEXT NOT NULL DEFAULT '';
+        "#
+    )?;
+    // Create a context row for every existing program and update the program row
+    let programs = ProgramRepository::new(connection).query_by_filter(ProgramFilter::new())?;
```

Comment:
I am starting at the previous version of the DB, so inserting the program like this should keep working. Or do I miss another part that might break?

> Another way to do this is to mark sync_buffer rows for master list as needing to be processed (I think nulling integration_date time on sync_buffer)

that sounds good! is there already some helper method for it? Otherwise, I have the feeling it should better be done in another PR...

Reply:
> I am starting at the previous version of the DB, so inserting the program like this should keep working. 

When we insert/update/select, diesel select/insert/update all fields, which may exist in repository but not in migrations yet. Btw i did make a suggestion but never shared: https://github.com/openmsupply/open-msupply/compare/1944-enrolment-encoutner-program-links...1944-enrolment-encoutner-program-links(Repository-free-migrations)

I started doing code changes below, however we can't call sync in the migrations. So can't really test or wait for sync integration before putting in a constraint. So not really applicable to this use case

```
use crate::StorageConnection;
use diesel::prelude::*;

table! {
    sync_buffer (record_id) {
        record_id -> Text,
        table_name -> Bool,
        integration_datetime -> Nullable<Timestamp>,
        integration_error -> Nullable<Text>,
    }
}

fn reset_sync_buffer_for_tables(
    table_names: Vec<&'static str>,
    connection: &StorageConnection,
) -> anyhow::Result<()> {
    diesel::update(sync_buffer::dsl::sync_buffer)
        .filter(sync_buffer::dsl::sync_buffer.eq_any(table_names))
        .set((
            sync_buffer::dsl::integration_datetime.eq(None),
            sync_buffer::dsl::integration_error.eq(None),
        ))
        .execute(&connection.connection)?;

    Ok(())
}
```

Reply:
great thanks! I was only looking at the test code and missed this... Will merge this into the latest PR (there are some conflicts already...)